### DISCUSSION TITLE
Unit tests and debugging. Inspect queries.

### DISCUSSION THREAD

**Discussion by espenhoh** [oracle/python-oracledb#129] [Q&A]
Hi!

I would like to stub out the database connection and be able to view and inspect the queries genereated by oracledb.
This is to enable unit tests and debugging query errors, without needing an actual oracle instance.

Diving through the source code of oracledb, I can't make sense of where the queries are actually generated.

I would be grateful if anybody have a suggestion on how to achieve this.

**Comment by anthony-tuininga** [comment]
I'm not sure exactly what you are trying to accomplish, but you can subclass the `Connection` and `Cursor` classes and then examine the arguments for functions you are interested in before passing them on to the parent class for actual execution. This can be seen in the [subclassing example](https://github.com/oracle/python-oracledb/blob/main/samples/subclassing.py).

I'm not sure what you mean by the phrase "queries are actually generated". The `oracledb` module doesn't generate any queries. It only executes whatever the developer specifically requests through `Cursor.execute()`. Can you clarify further?
